### PR TITLE
nixos/netbird: Allow running multiple netbird networks in parallel

### PR DESCRIPTION
**Pull request by Tom-Hubrecht**
## Description of changes

Update the netbird module to allow configuring multiple daemons through environment variable and to configure the interfaces used for the netbird network.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by misuzu** [comment]
What is the use-case for these new options? I don't really see the point of changing `NB_LOG_FILE`, `NB_CONFIG` or `interfaceName`, and the important stuff still [can't be declaratively configured](https://github.com/netbirdio/netbird/issues/430).

**Comment by Tom-Hubrecht** [comment]
> What is the use-case for these new options? I don't really see the point of changing `NB_LOG_FILE`, `NB_CONFIG` or `interfaceName`, and the important stuff still [can't be declaratively configured](https://github.com/netbirdio/netbird/issues/430). Also the `interfaceName` option won't actually change the network interface used by netbird.

Well I was under the impression with https://github.com/netbirdio/netbird/blob/main/client/cmd/root.go#L139 that it was possible to set a custom address for the management server and the admin endpoint. I agree that it isn't usefull to change `NB_CONFIG` and `NB_LOG_FILE`, but as I would set `NB_ADMIN_URL` for example in settings I thought it was logical to also set them here.
Hmmm, I'll delete the stuff about the interface then.

**Comment by misuzu** [comment]
> but as I would set `NB_ADMIN_URL` for example in settings I thought it was logical to also set them here

The official [documentation](https://docs.netbird.io/how-to/installation#running-net-bird-with-a-setup-key) suggests that it should be passed to the `netbird up` command:

> Alternatively, if you are hosting your own Management Service provide --management-url property pointing to your Management Service

I thing it better to just do that and provide examples in the documentation of your management module.

**Comment by Tom-Hubrecht** [comment]
@misuzu now that netbird (>= v0.25.4) allow configuring the interface and port of the wireguard tunnel, I think we can revisit this PR.
I also changed it so that we can have multiple netbird tunnels at the same time.

File: nixos/modules/services/networking/netbird.nix [hunk 1]
```diff
@@ -1,60 +1,169 @@
...
   options.services.netbird = {
     enable = mkEnableOption (lib.mdDoc "Netbird daemon");
     package = mkPackageOption pkgs "netbird" { };
-  };
-
-  config = mkIf cfg.enable {
-    boot.extraModulePackages = optional (versionOlder kernel.kernel.version "5.6") kernel.wireguard;
 
-    environment.systemPackages = [ cfg.package ];
+    tunnels = mkOption {
```
Comment (misuzu) [inline nixos/modules/services/networking/netbird.nix:46]: This probably needs a release note.
Also it's not clear how to actually use this option, what arguments should be passed to `netbird up`? Is `services.netbird.enabled` should be set? This kind of stuff deserves a documentation in `meta.doc`

File: nixos/modules/services/networking/netbird.nix [hunk 2]
```diff
@@ -1,60 +1,169 @@
...
+    ;
+
   kernel = config.boot.kernelPackages;
-  interfaceName = "wt0";
-in {
+
+  cfg = config.services.netbird;
+in
+{
   meta.maintainers = with maintainers; [ misuzu ];
```
Comment (misuzu) [inline nixos/modules/services/networking/netbird.nix:38]:
```suggestion
  meta.maintainers = with maintainers; [ misuzu thubrecht ];
```

File: nixos/modules/services/networking/netbird.nix [hunk 3]
```diff
@@ -1,60 +1,169 @@
...
 
+  disabledModules = [ "services/networking/netbird.nix" ];
```
Comment (misuzu) [inline nixos/modules/services/networking/netbird.nix:40]:
```suggestion

```

**Review by misuzu — COMMENTED** [review]
A few suggestions, but lgtm otherwise

**Review by misuzu — APPROVED** [review]

File: nixos/modules/services/networking/netbird.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+It is strictly equivalent to setting:
+
+```nix
+services.netbird.tunnels.wt0.stateDir = "netbird";
+```
+
+The `enable` option is mainly kept for backward compatibility, as defining netbird
+tunnels through the `tunnels` option is more expressive.
+
+## Multiple connexions setup {#module-services-netbird-multiple-connexions}
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird.md:23]:
```suggestion
## Multiple connections setup {#module-services-netbird-multiple-connections}
```

french spotted

File: nixos/modules/services/networking/netbird.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+netbird --daemon-addr unix:///var/run/netbird-wt1/sock ...
+```
+
+The address will by default be `unix:///var/run/netbird-<name>`.
+
+It is also possible to overwrite default options passed to the service, for
+example:
+
+```nix
+services.netbirs.tunnels.wt1.environment = {
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird.md:51]:
```suggestion
services.netbird.tunnels.wt1.environment = {
```

File: nixos/modules/services/networking/netbird.nix [hunk 6]
```diff
@@ -1,60 +1,168 @@
...
+      description = ''
+        Attribute set of Netbird tunnels, each one will spawn a daemon listening on ...
+      '';
     };
   };
+
+  config = mkMerge [
+    (mkIf cfg.enable {
+      # For backwards compatibility
+      services.netbird.tunnels.wt0.stateDir = "netbird";
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird.nix:104]: `lib.mkDefault` no on the value?

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Some typo stuff, but all good for me! We can merge once it's done.

File: nixos/modules/services/networking/netbird.nix (hunk 6 above)
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird.nix:104]: I mean, the `enable` option is only for backwards compat, if the user wants to change teh stateDir for `wt0`, then the way is to declare it through `tunnels.wt0`, and don't set the enable option

Comment (RaitoBezarius) [reply nixos/modules/services/networking/netbird.nix:107]: That works for me!

**Review by RaitoBezarius — APPROVED** [review]

**Comment by nazarewk** [comment]
nice, that's a perfect base I can put remainder of my personal configs to https://github.com/nazarewk-iac/nix-configs/blob/fb8bac011db7075b9c4f473ae4490a2ac727b29c/modules/networking/netbird/default.nix

**Pull request merged**